libnfc on macOS

libnfc from the cache does not seem to work for me:

~~~~
$ nix-shell -p libnfc

[nix-shell:~]$ nfc-list
nfc-list uses libnfc 1.7.1
Segmentation fault: 11
~~~~

What happens is a null-pointer dereference somewhere inside IOKit:

~~~~
* thread #1: tid = 0x4e52b7, 0x00007fff92a1bb69 libobjc.A.dylib`objc_msgSend + 41, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x00007fff92a1bb69 libobjc.A.dylib`objc_msgSend + 41
    frame #1: 0x00007fff7fd1700f IOKit`IORegistryEntrySearchCFProperty + 46
    frame #2: 0x00000001000c1bde libusb-1.0.0.dylib`process_new_device + 94
    frame #3: 0x00000001000bc68a libusb-1.0.0.dylib`darwin_init + 234
    frame #4: 0x00000001000b722f libusb-1.0.0.dylib`libusb_init + 559
    frame #5: 0x00000001000ae9cc libusb-0.1.4.dylib`usb_init + 28
    frame #6: 0x000000010009d5b0 libnfc.5.dylib`usb_prepare + 80
    frame #7: 0x000000010009db1e libnfc.5.dylib`acr122_usb_scan + 30
    frame #8: 0x0000000100091503 libnfc.5.dylib`nfc_list_devices + 563
    frame #9: 0x0000000100000ca4 nfc-list`main + 292
~~~~

I am really not sure what is going on, but what troubles me is that I cannot reproduce the build on my laptop: apparently, the examples depend on readline, which is not listed as a dependency.

~~~~
Making all in examples
make[2]: Entering directory '/private/tmp/libnfc-1.7.1/examples'
Making all in pn53x-tamashell-scripts
make[3]: Entering directory '/private/tmp/libnfc-1.7.1/examples/pn53x-tamashell-scripts'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/private/tmp/libnfc-1.7.1/examples/pn53x-tamashell-scripts'
make[3]: Entering directory '/private/tmp/libnfc-1.7.1/examples'
  CC     nfc-anticol.o
  CCLD   nfc-anticol
  CC     nfc-dep-initiator.o
  CCLD   nfc-dep-initiator
  CC     nfc-dep-target.o
  CCLD   nfc-dep-target
  CC     nfc-emulate-forum-tag2.o
  CCLD   nfc-emulate-forum-tag2
  CC     nfc-emulate-tag.o
  CCLD   nfc-emulate-tag
  CC     nfc-emulate-uid.o
  CCLD   nfc-emulate-uid
  CC     nfc-mfsetuid.o
  CCLD   nfc-mfsetuid
  CC     nfc-poll.o
  CCLD   nfc-poll
  CC     nfc-relay.o
  CCLD   nfc-relay
  CC     pn53x-diagnose.o
  CCLD   pn53x-diagnose
  CC     pn53x-sam.o
  CCLD   pn53x-sam
  CC     pn53x-tamashell.o
pn53x-tamashell.c:48:12: fatal error: 'readline/readline.h' file not found
#  include <readline/readline.h>
           ^
1 error generated.
make[3]: *** [Makefile:616: pn53x-tamashell.o] Error 1
make[3]: Leaving directory '/private/tmp/libnfc-1.7.1/examples'
make[2]: *** [Makefile:688: all-recursive] Error 1
make[2]: Leaving directory '/private/tmp/libnfc-1.7.1/examples'
make[1]: *** [Makefile:387: all-recursive] Error 1
make[1]: Leaving directory '/private/tmp/libnfc-1.7.1'
make: *** [Makefile:296: all] Error 2
~~~~

How does the [job](http://hydra.nixos.org/build/46783117) manage to succeed on Hydra? What’s wrong with libnfc or libusb built on Hydra?

After I added `readline` as a dependency the build succeded and the resulting libnfc is fully functional.

--- a/pkgs/build-support/setup-hooks/fix-darwin-frameworks.sh
+++ b/pkgs/build-support/setup-hooks/fix-darwin-frameworks.sh
@@ -1,0 +1,31 @@
+# On Mac OS X, frameworks are linked to the system CoreFoundation but
+# dynamic libraries built with nix use a pure version of CF this
+# causes segfaults for binaries that depend on it at runtime.  This
+# can be solved in two ways.
+# 1. Rewrite references to the pure CF using this setup hook, this
+# works for the simple case but this can still cause problems if other
+# dependencies (eg. python) use the pure CF.
+# 2. Create a wrapper for the binary that sets DYLD_FRAMEWORK_PATH to
+# /System/Library/Frameworks.  This will make everything load the
+# system's CoreFoundation framework while still keeping the
+# dependencies pure for other packages.
+
+fixupOutputHooks+=('fixDarwinFrameworksIn $prefix')
+
+fixDarwinFrameworks() {
+    local systemPrefix='/System/Library/Frameworks'
+
+    for fn in "$@"; do
+        if [ -L "$fn" ]; then continue; fi
+        echo "$fn: fixing dylib"
+
+        for framework in $(otool -L "$fn" | awk '/CoreFoundation\.framework/ {print $1}'); do
+          install_name_tool -change "$framework" "$systemPrefix/CoreFoundation.framework/Versions/A/CoreFoundation" "$fn" >&2
+        done
+    done
+}
+
+fixDarwinFrameworksIn() {
+    local dir="$1"
+    fixDarwinFrameworks $(find "$dir" -name "*.dylib")
+}

--- a/pkgs/os-specific/darwin/apple-sdk/default.nix
+++ b/pkgs/os-specific/darwin/apple-sdk/default.nix
@@ -119,6 +119,9 @@ let
 
     propagatedBuildInputs = deps;
 
+    # don't use pure CF for dylibs that depend on frameworks
+    setupHook = ../../../build-support/setup-hooks/fix-darwin-frameworks.sh;
+
     # allows building the symlink tree
     __impureHostDeps = [ "/System/Library/Frameworks/${name}.framework" ];
 

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -305,6 +305,8 @@ with pkgs;
 
   fixDarwinDylibNames = makeSetupHook { } ../build-support/setup-hooks/fix-darwin-dylib-names.sh;
 
+  fixDarwinFrameworks = makeSetupHook { } ../build-support/setup-hooks/fix-darwin-frameworks.sh;
+
   keepBuildTree = makeSetupHook { } ../build-support/setup-hooks/keep-build-tree.sh;
 
   enableGCOVInstrumentation = makeSetupHook { } ../build-support/setup-hooks/enable-coverage-instrumentation.sh;

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -8733,7 +8733,8 @@ with pkgs;
   libusb = callPackage ../development/libraries/libusb {};
 
   libusb1 = callPackage ../development/libraries/libusb1 {
-    inherit (darwin) libobjc IOKit;
+    inherit (darwin) libobjc;
+    inherit (darwin.apple_sdk.frameworks) IOKit;
   };
 
   libusbmuxd = callPackage ../development/libraries/libusbmuxd { };

